### ISSUE TITLE
implement kubernetes metadata enrichment

### ISSUE THREAD

**Issue by kortemik** [teragrep/dkr_01#9]
it is possible to fetch kubernetes metadata from kubelet which manages the corresponding node.

Example call to kubelet is shown bellow:

curl -k --key /var/lib/minikube/certs/apiserver-kubelet-client.key --cert /var/lib/minikube/certs/apiserver-kubelet-client.crt https://localhost:10250/pods 

```
{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {},
  "items": [
    {
      "metadata": {
        "name": "etcd-minikube",
        "namespace": "kube-system",
        "uid": "f5bb3d327095d4906ada206fb0e9a699",
        "creationTimestamp": null,
        "labels": {
          "component": "etcd",
          "tier": "control-plane"
        },
        "annotations": {
          "kubeadm.kubernetes.io/etcd.advertise-client-urls": "https://192.168.39.198:2379",
          "kubernetes.io/config.hash": "f5bb3d327095d4906ada206fb0e9a699",
          "kubernetes.io/config.seen": "2023-04-28T12:28:36.042764386Z",
          "kubernetes.io/config.source": "file"
        }
      },
      "spec": {
        "volumes": [
          {
            "name": "etcd-certs",
            "hostPath": {
              "path": "/var/lib/minikube/certs/etcd",
              "type": "DirectoryOrCreate"
            }
          },
          {
            "name": "etcd-data",
            "hostPath": {
              "path": "/var/lib/minikube/etcd",
              "type": "DirectoryOrCreate"
            }
          }
        ],
        "containers": [
          {
            "name": "etcd",
            "image": "k8s.gcr.io/etcd:3.5.3-0",
            "command": [
              "etcd",
              "--advertise-client-urls=https://192.168.39.198:2379",
              "--cert-file=/var/lib/minikube/certs/etcd/server.crt",
              "--client-cert-auth=true",
              "--data-dir=/var/lib/minikube/etcd",
              "--experimental-initial-corrupt-check=true",
              "--initial-advertise-peer-urls=https://192.168.39.198:2380",
              "--initial-cluster=minikube=https://192.168.39.198:2380",
              "--key-file=/var/lib/minikube/certs/etcd/server.key",
              "--listen-client-urls=https://127.0.0.1:2379,https://192.168.39.198:2379",
              "--listen-metrics-urls=http://127.0.0.1:2381",
              "--listen-peer-urls=https://192.168.39.198:2380",
              "--name=minikube",
              "--peer-cert-file=/var/lib/minikube/certs/etcd/peer.crt",
              "--peer-client-cert-auth=true",
              "--peer-key-file=/var/lib/minikube/certs/etcd/peer.key",
              "--peer-trusted-ca-file=/var/lib/minikube/certs/etcd/ca.crt",
              "--proxy-refresh-interval=70000",
              "--snapshot-count=10000",
              "--trusted-ca-file=/var/lib/minikube/certs/etcd/ca.crt"
            ],
            "resources": {
              "requests": {
                "cpu": "100m",
                "memory": "100Mi"
              }
            },
            "volumeMounts": [
              {
                "name": "etcd-data",
                "mountPath": "/var/lib/minikube/etcd"
              },
              {
                "name": "etcd-certs",
                "mountPath": "/var/lib/minikube/certs/etcd"
              }
            ],
            "livenessProbe": {
              "httpGet": {
                "path": "/health",
                "port": 2381,
                "host": "127.0.0.1",
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 10,
              "timeoutSeconds": 15,
              "periodSeconds": 10,
              "successThreshold": 1,
              "failureThreshold": 8
            },
            "startupProbe": {
              "httpGet": {
                "path": "/health",
                "port": 2381,
                "host": "127.0.0.1",
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 10,
              "timeoutSeconds": 15,
              "periodSeconds": 10,
              "successThreshold": 1,
              "failureThreshold": 24
            },
            "terminationMessagePath": "/dev/termination-log",
            "terminationMessagePolicy": "File",
            "imagePullPolicy": "IfNotPresent"
          }
        ],
        "restartPolicy": "Always",
        "terminationGracePeriodSeconds": 30,
        "dnsPolicy": "ClusterFirst",
        "nodeName": "minikube",
        "hostNetwork": true,
        "securityContext": {
          "seccompProfile": {
            "type": "RuntimeDefault"
          }
        },
        "schedulerName": "default-scheduler",
        "tolerations": [
          {
            "operator": "Exists",
            "effect": "NoExecute"
          }
        ],
        "priorityClassName": "system-node-critical",
        "enableServiceLinks": true
      },
      "status": {
        "phase": "Running",
        "conditions": [
          {
            "type": "Initialized",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2023-04-28T12:28:36Z"
          },
          {
            "type": "Ready",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2023-04-28T12:28:54Z"
          },
          {
            "type": "ContainersReady",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2023-04-28T12:28:54Z"
          },
          {
            "type": "PodScheduled",
            "status": "True",
            "lastProbeTime": null,
            "lastTransitionTime": "2023-04-28T12:28:36Z"
          }
        ],
        "hostIP": "192.168.39.198",
        "podIP": "192.168.39.198",
        "podIPs": [
          {
            "ip": "192.168.39.198"
          }
        ],
        "startTime": "2023-04-28T12:28:36Z",
        "containerStatuses": [
          {
            "name": "etcd",
            "state": {
              "running": {
                "startedAt": "2023-04-28T12:28:38Z"
              }
            },
            "lastState": {
              "terminated": {
                "exitCode": 255,
                "reason": "Error",
                "startedAt": "2023-03-30T14:29:21Z",
                "finishedAt": "2023-04-28T12:28:25Z",
                "containerID": "docker://dfeb0a1ff0fa506a00d9170f764fb54bc77070ef5154fa0281b96db1d63e6baf"
              }
            },
            "ready": true,
            "restartCount": 4,
            "image": "k8s.gcr.io/etcd:3.5.3-0",
            "imageID": "docker-pullable://k8s.gcr.io/etcd@sha256:13f53ed1d91e2e11aac476ee9a0269fdda6cc4874eba903efd40daf50c55eee5",
            "containerID": "docker://92703e3ddeb25a2677af4daf44bd54ea42926a1483916e6fc4b682ece0c5fb65",
            "started": true
          }
        ],
        "qosClass": "Burstable"
      }
    }
  ]
}
```

logging driver is already aware of the containerId.

containerId can be used to extract further meta information from the kubelet response.
```
jq '.items[].status.containerStatuses[].containerID' kubelet.json
"docker://92703e3ddeb25a2677af4daf44bd54ea42926a1483916e6fc4b682ece0c5fb65"

jq '.items[].metadata.name' kubelet.json
"etcd-minikube"

jq '.items[].metadata.namespace' kubelet.json
"kube-system"

jq '.items[].metadata.labels' kubelet.json
{
  "component": "etcd",
  "tier": "control-plane"
}

jq '.items[].metadata.uid' kubelet.json
"f5bb3d327095d4906ada206fb0e9a699"

jq '.items[].metadata.creationTimestamp' kubelet.json
null
```

please extract items above.

please make following configureable:
* k8s metadata extraction enabled (toggles the whole feature)
* * kubelet address (i.e. https://localhost:10250)
* * client auth enabled (toggles if client cert and key are to be used for authentication)
* * * tls-client-auth certificate location
* * * tls-client-auth key location
* * server certificate validation (toggles if server certificate is validated)
* * * tls-connection ca-certificate (server cafile to check server certificate against)
* * metadata refresh enabled
* * * metadata rerfresh interval (seconds, 0 means always, for each request)

further information will be provided what to do with the extracted information.

store the information on per container basis (into the logPair relation). it must not be refreshed if data is present unless refresh is enabled and it has expired.

**Comment by kortemik** [comment]
@StrongestNumber9 please check above validity as you are the developer of k8s_01

**Comment by eemhu** [comment]
implemented in PR #11 

needs testing via minikube, will make another ticket about it #19
